kythe/rust/indexer/src/indexer/analyzers.rs

```diff
@@ -194,6 +286,101 @@ impl<'a, 'b> CrateAnalyzer<'a, 'b> {
         Ok(())
     }
 
+    /// Emits tbuiltin nodes for all of the Rust built-in types
+    pub fn emit_tbuiltin_nodes(&mut self) -> Result<(), KytheError> {
+        for vname in self.type_vnames.values() {
+            self.emitter.emit_node(vname, "/kythe/node/kind", b"tbuiltin".to_vec())?;
+        }
+
+        Ok(())
+    }
+
+    /// Creates the internal `method_index` by analyzing the implementations and
+    /// relations on the save_analysis.
+    ///
+    /// Must be called before "emit_definitions"
+    pub fn process_implementations(&mut self) -> Result<(), KytheError> {
+        // Create a HashMap mapping the implementation Id to the implementation
+        // It might be a safe assumption that the index is the Id, but we can't be too
+        // careful
+        let impls = self.krate.analysis.impls.clone();
+        let mut impl_map: HashMap<u32, rls_data::Impl> = HashMap::new();
+        for implementation in impls.iter() {
+            impl_map.insert(implementation.id, implementation.clone());
+        }
+        // Drop the cloned vector to save memory
+        drop(impls);
+
+        // Create a HashMap betwen a method definition Id and the struct and trait being
+        // implemented on
+        let mut method_index: HashMap<rls_data::Id, MethodImpl> = HashMap::new();
+        let relations = &self.krate.analysis.relations;
+        for relation in relations.iter() {
+            // If this is an implementation relation
+            if let rls_data::RelationKind::Impl { id: impl_id, .. } = relation.kind {
+                let implementation = impl_map.get(&impl_id).ok_or_else(|| {
+                    KytheError::IndexerError(format!(
+                        "Couldn't find implementation for relation {:?}",
+                        relation
+                    ))
+                })?;
+                // Add all of the childred to the HashMap
+                for child in implementation.children.iter() {
+                    // The struct being implemented on
+                    let struct_target = relation.from;
+                    // The optional trait being implemented on
+                    let trait_target =
+                        if relation.to.krate != 4294967295 { Some(relation.to) } else { None };
```

Comment:
Is there a clearer way to represent this? Can you use -1 or ~0?

Reply:
This maxint is the from the data we are given in the save_analysis

Reply:
Granted that's the case, but it's a weird constant to run into when reading the code (if you're not familiar with the vagaries of 2's compliment arithmetic). Could you give it a name?